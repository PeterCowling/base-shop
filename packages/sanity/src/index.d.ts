import "server-only";
import type { SanityBlogConfig } from "@acme/types";
export interface ProductBlock {
    _type: "productReference";
    slug: string;
}
export interface BlogPost {
    title: string;
    slug: string;
    excerpt?: string;
    body?: (unknown | ProductBlock)[];
    mainImage?: string;
    author?: string;
    categories?: string[];
    products?: string[];
}
export declare function getConfig(shopId: string): Promise<SanityBlogConfig>;
export declare function fetchPublishedPosts(shopId: string): Promise<BlogPost[]>;
export declare function fetchPostBySlug(shopId: string, slug: string): Promise<BlogPost | null>;
export declare function publishQueuedPost(shopId: string): Promise<void>;
