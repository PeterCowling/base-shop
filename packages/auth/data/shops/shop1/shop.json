{
  "id": "shop1",
  "name": "shop1",
  "catalogFilters": [],
  "themeId": "base",
  "themeDefaults": {},
  "themeOverrides": {},
  "themeTokens": {},
  "filterMappings": {},
  "priceOverrides": {},
  "localeOverrides": {},
  "navigation": [],
  "analyticsEnabled": false,
  "shippingProviders": [],
  "taxProviders": [
    "taxjar"
  ],
  "paymentProviders": [],
  "enableEditorial": false,
  "subscriptionsEnabled": false,
  "rentalSubscriptions": []
}