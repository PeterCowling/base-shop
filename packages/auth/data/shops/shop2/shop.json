{
  "id": "shop2",
  "name": "shop2",
  "catalogFilters": [],
  "themeId": "base",
  "themeDefaults": {},
  "themeOverrides": {},
  "themeTokens": {},
  "filterMappings": {},
  "priceOverrides": {},
  "localeOverrides": {},
  "navigation": [
    {
      "label": "Parent",
      "url": "/parent",
      "children": [
        {
          "label": "Child",
          "url": "/child"
        }
      ]
    }
  ],
  "analyticsEnabled": false,
  "shippingProviders": [],
  "taxProviders": [
    "taxjar"
  ],
  "paymentProviders": [],
  "enableEditorial": false,
  "subscriptionsEnabled": false,
  "rentalSubscriptions": []
}