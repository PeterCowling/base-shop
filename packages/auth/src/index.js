// packages/auth/src/index.ts
export { canRead, canWrite, READ_ROLES, WRITE_ROLES } from "./rbac";
export { extendRoles, isRole } from "./types/roles";
export { hasPermission } from "./permissions";
export { requirePermission } from "./requirePermission";
export { CUSTOMER_SESSION_COOKIE, CSRF_TOKEN_COOKIE, getCustomerSession, createCustomerSession, destroyCustomerSession, listSessions, revokeSession, validateCsrfToken, } from "./session";
export { setSessionStoreFactory } from "./store";
export { enrollMfa, verifyMfa, isMfaEnabled, } from "./mfa";
