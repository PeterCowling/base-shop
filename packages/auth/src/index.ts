export { canRead, canWrite, READ_ROLES, WRITE_ROLES } from "./rbac";
export type { Role } from "./types";
