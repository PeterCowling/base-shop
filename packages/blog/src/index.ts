import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const DATA_FILE = path.join(
  fileURLToPath(new URL("../../..", import.meta.url)),
  "data",
  "blog",
  "posts.json",
);

export interface BlogPost {
  id: string;
  title: string;
  slug: string;
  excerpt?: string;
  content?: string;
  skus: string[];
  publishedAt?: string;
  scheduledAt?: string;
}

function readPosts(): BlogPost[] {
  try {
    const raw = readFileSync(DATA_FILE, "utf8");
    return JSON.parse(raw) as BlogPost[];
  } catch {
    return [];
  }
}

function writePosts(posts: BlogPost[]): void {
  writeFileSync(DATA_FILE, JSON.stringify(posts, null, 2));
}

export function createPost(post: BlogPost): void {
  const posts = readPosts();
  posts.push(post);
  writePosts(posts);
}

export function publishScheduledPosts(now: Date = new Date()): void {
  const posts = readPosts();
  let changed = false;
  for (const p of posts) {
    if (
      p.scheduledAt &&
      !p.publishedAt &&
      new Date(p.scheduledAt).getTime() <= now.getTime()
    ) {
      p.publishedAt = now.toISOString();
      changed = true;
    }
  }
  if (changed) writePosts(posts);
}

export function getPublishedPosts(now: Date = new Date()): BlogPost[] {
  publishScheduledPosts(now);
  return readPosts().filter((p) => {
    if (p.publishedAt) return new Date(p.publishedAt).getTime() <= now.getTime();
    return false;
  });
}

export function getPostBySlug(
  slug: string,
  now: Date = new Date(),
): BlogPost | undefined {
  return getPublishedPosts(now).find((p) => p.slug === slug);
}
