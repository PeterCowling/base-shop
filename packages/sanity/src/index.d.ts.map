{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,OAAO,aAAa,CAAC;AAIrB,OAAO,EAA0B,KAAK,gBAAgB,EAAE,MAAM,aAAa,CAAC;AAG5E,MAAM,WAAW,aAAa;IAC5B,KAAK,EAAE,MAAM,CAAC;IACd,CAAC,GAAG,EAAE,MAAM,GAAG,OAAO,CAAC;CACxB;AAED,MAAM,WAAW,YAAa,SAAQ,aAAa;IACjD,KAAK,EAAE,kBAAkB,CAAC;IAC1B,IAAI,EAAE,MAAM,CAAC;CACd;AAED,MAAM,WAAW,QAAQ;IACvB,KAAK,EAAE,MAAM,CAAC;IACd,IAAI,EAAE,MAAM,CAAC;IACb,OAAO,CAAC,EAAE,MAAM,CAAC;IACjB,IAAI,CAAC,EAAE,aAAa,EAAE,CAAC;IACvB,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB,MAAM,CAAC,EAAE,MAAM,CAAC;IAChB,UAAU,CAAC,EAAE,MAAM,EAAE,CAAC;IACtB,QAAQ,CAAC,EAAE,MAAM,EAAE,CAAC;CACrB;AAED,wBAAsB,SAAS,CAAC,MAAM,EAAE,MAAM,GAAG,OAAO,CAAC,gBAAgB,CAAC,CAOzE;AAaD,wBAAsB,mBAAmB,CAAC,MAAM,EAAE,MAAM,GAAG,OAAO,CAAC,QAAQ,EAAE,CAAC,CAS7E;AAED,wBAAsB,eAAe,CACnC,MAAM,EAAE,MAAM,EACd,IAAI,EAAE,MAAM,GACX,OAAO,CAAC,QAAQ,GAAG,IAAI,CAAC,CAS1B;AAED,wBAAsB,iBAAiB,CAAC,MAAM,EAAE,MAAM,GAAG,OAAO,CAAC,IAAI,CAAC,CAcrE"}