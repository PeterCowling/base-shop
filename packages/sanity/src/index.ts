import "server-only";
import { createClient } from "@sanity/client";
import { getSanityConfig } from "@platform-core/shops";
import { getShopById } from "@platform-core/repositories/shop.server";
import type { SanityBlogConfig } from "@acme/types";
import { nowIso } from "@date-utils";

export interface ProductBlock {
  _type: "productReference";
  slug: string;
}

export interface BlogPost {
  title: string;
  slug: string;
  excerpt?: string;
  body?: (unknown | ProductBlock)[];
  mainImage?: string;
  author?: string;
  categories?: string[];
  products?: string[];
}

export async function getConfig(shopId: string): Promise<SanityBlogConfig> {
  const shop = await getShopById(shopId);
  const config = getSanityConfig(shop);
  if (!config) {
    throw new Error(`Missing Sanity credentials for shop ${shopId}`);
  }
  return config as any;
}

async function getClient(shopId: string) {
  const { projectId, dataset, token: sanityToken } = await getConfig(shopId);
  return createClient({
    projectId,
    dataset,
    token: sanityToken,
    apiVersion: "2023-01-01",
    useCdn: true,
  });
}

export async function fetchPublishedPosts(shopId: string): Promise<BlogPost[]> {
  try {
    const client = await getClient(shopId);
    const query = `*[_type == "post" && defined(slug.current) && published == true && !(_id in path('drafts.**'))]{title, "slug": slug.current, excerpt, mainImage, author, categories, products}`;
    const posts = await client.fetch<BlogPost[]>(query);
    return posts;
  } catch {
    return [];
  }
}

export async function fetchPostBySlug(
  shopId: string,
  slug: string,
): Promise<BlogPost | null> {
  try {
    const client = await getClient(shopId);
    const query = `*[_type == "post" && slug.current == $slug && published == true][0]{title, "slug": slug.current, excerpt, mainImage, author, categories, products, body[]{..., _type == "productReference" => { _type, slug }}}`;
    const post = await client.fetch<BlogPost | null>(query, { slug });
    return post;
  } catch {
    return null;
  }
}

export async function publishQueuedPost(shopId: string): Promise<void> {
  try {
    const client = await getClient(shopId);
    const next = await client.fetch<{ _id: string } | null>(
      `*[_type == "post" && published != true && !(_id in path('drafts.**'))]|order(_createdAt asc)[0]{_id}`,
    );
    if (!next?._id) return;
    await client
      .patch(next._id)
      .set({ published: true, publishedAt: nowIso() })
      .commit();
  } catch {
    // swallow errors; scheduled job will handle logging
  }
}
