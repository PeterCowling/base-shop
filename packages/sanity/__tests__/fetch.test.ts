jest.mock('@sanity/client', () => ({
  createClient: jest.fn(),
}));

jest.mock('@platform-core/repositories/shop.server', () => ({
  getShopById: jest.fn(),
}));

jest.mock('@platform-core/shops', () => ({
  getSanityConfig: jest.fn(),
}));

import { fetchPublishedPosts, fetchPostBySlug } from '../src';
import { createClient } from '@sanity/client';
import { getShopById } from '@platform-core/repositories/shop.server';
import { getSanityConfig } from '@platform-core/shops';

describe('fetch helpers', () => {
  const createClientMock = createClient as jest.Mock;
  const getShopByIdMock = getShopById as jest.Mock;
  const getSanityConfigMock = getSanityConfig as jest.Mock;
  let fetchMock: jest.Mock;

  beforeEach(() => {
    fetchMock = jest.fn();
    createClientMock.mockReturnValue({ fetch: fetchMock });
    getShopByIdMock.mockResolvedValue({ id: 'shop1' });
    getSanityConfigMock.mockReturnValue({
      projectId: 'pid',
      dataset: 'ds',
      token: 'tkn',
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('fetchPublishedPosts returns posts on success', async () => {
    const posts = [{ title: 'Post', slug: 'post' }];
    fetchMock.mockResolvedValue(posts);
    await expect(fetchPublishedPosts('shop1')).resolves.toEqual(posts);
    expect(fetchMock.mock.calls[0][0]).toMatchSnapshot();
  });

  it('fetchPublishedPosts returns empty array on failure', async () => {
    fetchMock.mockRejectedValue(new Error('fail'));
    await expect(fetchPublishedPosts('shop1')).resolves.toEqual([]);
  });

  it('fetchPostBySlug returns post on success', async () => {
    const post = { title: 'Post', slug: 'post' };
    fetchMock.mockResolvedValue(post);
    await expect(fetchPostBySlug('shop1', 'post')).resolves.toEqual(post);
    expect(fetchMock.mock.calls[0][0]).toMatchSnapshot();
    expect(fetchMock.mock.calls[0][1]).toEqual({ slug: 'post' });
  });

  it('fetchPostBySlug returns null on failure', async () => {
    fetchMock.mockRejectedValue(new Error('fail'));
    await expect(fetchPostBySlug('shop1', 'post')).resolves.toBeNull();
  });
});

