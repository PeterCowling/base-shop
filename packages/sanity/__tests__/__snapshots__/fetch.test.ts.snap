// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch helpers fetchPublishedPosts returns posts on success 1`] = `"*[_type == \"post\" && defined(slug.current) && published == true && !(_id in path('drafts.**'))]{title, \"slug\": slug.current, excerpt, mainImage, author, categories, products}"`;

exports[`fetch helpers fetchPostBySlug returns post on success 1`] = `"*[_type == \"post\" && slug.current == $slug && published == true][0]{title, \"slug\": slug.current, excerpt, mainImage, author, categories, products, body[]{..., _type == \"productReference\" => { _type, slug }}}"`;
